source/loader/layers/validation/ur_valddi.cpp

```diff
@@ -9,7 +9,51 @@
  */
 #include "ur_validation_layer.hpp"
 
+#include <atomic>
+#include <functional>
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+#include <utility>
+
 namespace validation_layer {
+struct {
+    std::map<void *, std::atomic<int>> counts;
+    std::shared_timed_mutex mutex;
+
+    std::pair<void *, int> incrementRefCount(void *ptr) {
+        std::shared_lock<std::shared_timed_mutex> lock(mutex);
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            lock.unlock();
+            std::unique_lock<std::shared_timed_mutex> ulock(mutex);
+            it->second.fetch_add(1, std::memory_order_relaxed);
+            return std::make_pair(it->first, it->second.load());
+        } else {
+            counts.emplace(ptr, 1);
+            return std::make_pair(ptr, 1);
+        }
+    }
+
+    std::pair<void *, int> decrementRefCount(void *ptr) {
+        std::shared_lock<std::shared_timed_mutex> lock(mutex);
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            if (it->second.fetch_sub(1, std::memory_order_release) == 1) {
+                lock.unlock();
+                std::unique_lock<std::shared_timed_mutex> ulock(mutex);
```

Comment:
Race here, similarly as in `incrementRefCount`

Reply:
Should be resolved now.